bundle/edu.gemini.qpt.client/src/main/java/edu/gemini/qpt/core/listeners/LimitsListener.java

```diff
@@ -103,14 +100,26 @@ public void propertyChange(PropertyChangeEvent evt) {
             final Interval nightInterval   = new Interval(tbn.getStartTime(), tbn.getEndTime());
             final Supplier<Union<Interval>> wholeNight = () -> new Union<>(nightInterval);
 
-            final Predicate<String> contributes =
-                (cacheName) -> {
-                    final Map<Obs, Union<Interval>> c = v.getSchedule().getCache(cacheName);
-                    final Union<Interval> u = c.getOrDefault(a.getObs(), new Union<>());
-
-                    final Union<Interval> r = wholeNight.get();
-                    r.remove(u);
-                    return !r.isEmpty();
+            final BiFunction<Function<Interval, Long>, String, Predicate<Long>> matchWith =
+                (f, cacheName) ->
+                    ts -> {
+                        final Map<Obs, Union<Interval>> c = v.getSchedule().getCache(cacheName);
+                        final Union<Interval> u = c.getOrDefault(a.getObs(), new Union<>());
+
+                        return u.getIntervals().stream().anyMatch(i -> f.apply(i) == ts);
+                    };
+
+            final BiFunction<Function<Interval, Long>, Interval, Option<String>> explainLimit =
+                (f, i) -> {
+                    if (matchWith.apply(f, Variant.VISIBLE_UNION_CACHE).test(f.apply(i))) {
+                        return new Some<>("%s(A)");
```

Comment:
Another issue that I just happened to spot is that they requested an "A" for airmass limits and an "H" for hour angle. It all boils down to elevation constraints and the `VISIBLE_UNION_CACHE`, but we should try to show the appropriate letter for the observation.  We know the constraint that should be used from the observation via `obs.getElevationConstraintType()`.  So for example:

```java
final String flag;
switch (obs.getElevationConstraintType()) {
    case HOUR_ANGLE:
        flag = "H";
        break;
    default:
        flag = "A";
        break;
}
```

The hard part will be shuffling around the arguments to `explainLimit`.  Let's update the formatting of the timestamps and then look into this issue.